### PR TITLE
Add LinkedIn provider

### PR DESCRIPTION
**Pull request by jerriep**

### CODE REVIEW

File: src/AspNet.Security.OAuth.LinkedIn/LinkedInAuthenticationHandler.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            var response = await Backchannel.SendAsync(request, Context.RequestAborted);
+            response.EnsureSuccessStatusCode();
+
+            var payload = JObject.Parse(await response.Content.ReadAsStringAsync());
+
+            identity.AddOptionalClaim(ClaimTypes.NameIdentifier, LinkedInAuthenticationHelper.GetIdentifier(payload), Options.ClaimsIssuer)
+                .AddOptionalClaim(ClaimTypes.Name, LinkedInAuthenticationHelper.GetName(payload), Options.ClaimsIssuer)
+                .AddOptionalClaim(ClaimTypes.Email, LinkedInAuthenticationHelper.GetEmail(payload), Options.ClaimsIssuer)
+                .AddOptionalClaim("urn:linkedin:profile", LinkedInAuthenticationHelper.GetPublicProfileUrl(payload), Options.ClaimsIssuer)
+                .AddOptionalClaim("urn:linkedin:profilepicture", LinkedInAuthenticationHelper.GetProfilePictureUrl(payload), Options.ClaimsIssuer);
```
Comment (jerriep) [inline src/AspNet.Security.OAuth.LinkedIn/LinkedInAuthenticationHandler.cs]: @PinpointTownes Not sure how you want to handle Claim names which are essentially two words, like "profile picture":

1. Do it all lowercase and one word, i.e. "profilepicture"
2. Camel cased, i.e. "profilePicture"
3. Add hyphen, i.e. "profile-picture"

Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.LinkedIn/LinkedInAuthenticationHandler.cs]: No personal preference :smile: 

Existing `ClaimTypes.*` claims always use the first one, so I'd be tempted to choose this option for consistency :+1:

File: AspNet.Security.OAuth.Providers.sln.DotSettings [hunk 2]
```diff
@@ -21,4 +21,8 @@
     <s:Boolean x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/SPACE_BEFORE_TYPEOF_PARENTHESES/@EntryValue">False</s:Boolean>
     <s:String x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/TYPE_DECLARATION_BRACES/@EntryValue">END_OF_LINE</s:String>
     <s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=PrivateInstanceFields/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="aaBb" /&gt;</s:String>
-    <s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=PrivateStaticFields/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="aaBb" /&gt;</s:String></wpf:ResourceDictionary>
\ No newline at end of file
+    <s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=PrivateStaticFields/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="aaBb" /&gt;</s:String>
```
Comment (PinpointTownes) [inline AspNet.Security.OAuth.Providers.sln.DotSettings]: These changes look unrelated. Any chance you could run `git rebase -i HEAD~5` and remove the first two commits?

File: samples/Mvc.Client/Startup.cs [hunk 3]
```diff
@@ -49,6 +49,11 @@ public class Startup {
                 options.ClientSecret = "98f1bf028608901e9df91d64ee61536fe562064b";
             });
 
+            app.UseLinkedInAuthentication(options => {
```
Comment (PinpointTownes) [inline samples/Mvc.Client/Startup.cs]: I guess you can't easily re-generate your client_id/client_secret via LinkedIn?
I'll try to create a register a new app to test it :sake:

File: src/AspNet.Security.OAuth.LinkedIn/LinkedInAuthenticationExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+using Microsoft.Framework.DependencyInjection;
+using Microsoft.Framework.Internal;
+using Microsoft.Framework.OptionsModel;
+
+namespace Microsoft.AspNet.Builder {
+    public static class LinkedInAuthenticationExtensions {
+        public static IServiceCollection ConfigureLinkedInAuthentication(
+            [NotNull] this IServiceCollection services,
+            [NotNull] Action<LinkedInAuthenticationOptions> configuration)
+        {
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.LinkedIn/LinkedInAuthenticationExtensions.cs]: I'm not sure why braces style is so inconsistent. I probably missed something when generating the `.DotSettings` file.

File: src/AspNet.Security.OAuth.LinkedIn/LinkedInAuthenticationHandler.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+            if (notification.Principal?.Identity == null)
+            {
+                return null;
+            }
+
+            return new AuthenticationTicket(
+                notification.Principal, notification.Properties,
+                notification.Options.AuthenticationScheme);
+
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.LinkedIn/LinkedInAuthenticationHandler.cs]: Same thing for the blank lines at the end for these classes... :cry:

File: src/AspNet.Security.OAuth.LinkedIn/project.json [hunk 6]
```diff
@@ -0,0 +1,20 @@
+﻿{
+    "version": "1.0.0-alpha1-*",
+    "description": "ASP.NET 5 security middleware enabling LinkedIn authentication.",
+    "authors": [ "Kévin Chalet", "Jerrie Pelser" ],
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.LinkedIn/project.json]: Since you're the only author of this great provider, you should probably add just your name :palm_tree:

**Comment by PinpointTownes** [comment]
Nice job, thanks @jerriep! :smile: 

nit: don't forget to run `git fetch upstream` + `git rebase upstream/dev` to get rid of the annoying `Merge remote-tracking branch 'aspnet-contrib/dev' into linkedin` commit :+1:

File: samples/Mvc.Client/Startup.cs (hunk 3 above)
Comment (jerriep) [inline samples/Mvc.Client/Startup.cs]: Yeah sorry, I delete it. I can recreate one for you to test if you want to. Let me know.

File: src/AspNet.Security.OAuth.LinkedIn/LinkedInAuthenticationExtensions.cs (hunk 4 above)
Comment (jerriep) [inline src/AspNet.Security.OAuth.LinkedIn/LinkedInAuthenticationExtensions.cs]: We'll get it right eventually... :)

File: AspNet.Security.OAuth.Providers.sln.DotSettings (hunk 2 above)
Comment (jerriep) [inline AspNet.Security.OAuth.Providers.sln.DotSettings]: Honestly not sure how to do that....? I get into some state that I do not know how to get out of. Afraid I may just mess things up even more

File: src/AspNet.Security.OAuth.LinkedIn/LinkedInAuthenticationHandler.cs (hunk 5 above)
Comment (jerriep) [inline src/AspNet.Security.OAuth.LinkedIn/LinkedInAuthenticationHandler.cs]: Should there be a blank line at the end?

File: AspNet.Security.OAuth.Providers.sln.DotSettings (hunk 2 above)
Comment (PinpointTownes) [inline AspNet.Security.OAuth.Providers.sln.DotSettings]: Don't worry about messing everything up: if things go really bad locally, just run `git reset --hard origin/linkedin` and you'll get back the current history of your fork on GitHub :smile: 

`git rebase -i` is rather easy to use, as it's graphical and uses your favorite text editor to edit the git history. When running `git rebase -i HEAD~5` (that takes the last 5 commits), select the line corresponding to the 2 commits you want to remove and delete them. Then, save the file and close your text editor. If you're happy with the changes (don't hesitate to use GitHub client's graphical history for that), then, just force push: `git push -f`.

Anyway, if you're afraid of doing bad things, I'll fix everything when merging :+1:

File: samples/Mvc.Client/Startup.cs (hunk 3 above)
Comment (PinpointTownes) [inline samples/Mvc.Client/Startup.cs]: That would be great, thanks! :smile:

File: src/AspNet.Security.OAuth.LinkedIn/LinkedInAuthenticationHandler.cs (hunk 5 above)
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.LinkedIn/LinkedInAuthenticationHandler.cs]: I'm not against having a final blank line at the end of the file, but here, these blank lines are just pointless.

File: AspNet.Security.OAuth.Providers.sln.DotSettings (hunk 2 above)
Comment (jerriep) [inline AspNet.Security.OAuth.Providers.sln.DotSettings]: I did as you suggested, but then the history looks completely funny to me. Maybe at a time when we are both online, we can get on a chat and I can show you how it looks on my side. Maybe even let you remote into my machine?

Damn, I need to you help me put a course together to help developers get their minds around these things. I sort of know Git and find it extremely frustrating and confusing. I can just imagine that this is how most developers must feel.

File: samples/Mvc.Client/Startup.cs (hunk 3 above)
Comment (jerriep) [inline samples/Mvc.Client/Startup.cs]: OK, when I get all the other issues fixed, I can commit with Client id and secret

File: AspNet.Security.OAuth.Providers.sln.DotSettings (hunk 2 above)
Comment (PinpointTownes) [inline AspNet.Security.OAuth.Providers.sln.DotSettings]: Yeah, git is/can be extremely frustrating. A nice `aspnetcast` video could be very helpful :smile:

**Comment by jerriep** [comment]
@PinpointTownes I have fix code style issues.

I have also added app key and secret. Callback URLs are set up for the ports on either the "IIS Express" or "web" profiles, so test under either of those.

**Comment by PinpointTownes** [comment]
Perfect, it works like a charm! :smile: 

Do you wanna try squashing your PR to have only one commit?
(I can do it locally if you prefer)

**Comment by jerriep** [comment]
Haha, if you can do it, it would be helpful... :blush: 

And also comment the command here, so I can add it to my course and know what to do next time :wink:

**Comment by PinpointTownes** [comment]
Merged, thanks: https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/commit/8c36e209354742fbb7e7eb6ba812fa7b4bb2137b :+1: 

Squashing is really easy when using interactive rebasing:

`git rebase -i HEAD~[number of commits added by your PR]`

It opens your default text editor and displays the list of commits. Replace `pick` by `f` (or `fixup`) each time you wanna "squash" a commit with the commit just above, save the file and close the editor. Use `git log` or GitHub Desktop to ensure the history looks fine and push :+1:

**Comment by jerriep** [comment]
Awesome stuff, thanks @PinpointTownes
